core/utils/exact_match_table.h

```diff
@@ -0,0 +1,440 @@
+class ExactMatchKeyHash {
+ public:
+  explicit ExactMatchKeyHash(size_t len) : len_(len) {}
+
+  HashResult operator()(const ExactMatchKey &key) const {
+    HashResult init_val = 0;
+
+    promise(len_ >= sizeof(uint64_t));
+    promise(len_ <= sizeof(ExactMatchKey));
+
+#if __SSE4_2__ && __x86_64
+    for (size_t i = 0; i < len_ / 8; i++) {
+      init_val = crc32c_sse42_u64(key.u64_arr[i], init_val);
+    }
+    return init_val;
+#else
+    return rte_hash_crc(&key, len_, init_val);
+#endif
+  }
+
+ private:
+  size_t len_;
+};
+
+// ExactMatchField describes a field to be matched on.
+// These should only be manually created when calling
+// ExactMatchTable::AddField(). In this case, only the `mask`, `offset`, and
+// `size` fields will be considered. `pos` and `attr_id` are set by
+// ExactMatchTable::AddField().
+struct ExactMatchField {
+  // bits with 1: the bit must be considered.
+  // bits with 0: don't care
+  uint64_t mask;
+
+  int attr_id;  // -1 for offset-based fields
+
+  // Relative offset in the packet data for offset-based fields.
+  // (starts from data_off, not the beginning of the headroom
+  int offset;
+
+  int pos;  // relative position in the key
+
+  int size;  // in bytes. 1 <= size <= MAX_FIELD_SIZE
+};
+
+static_assert(std::is_pod<ExactMatchField>::value,
+              "ExactMatchField is not a POD type");
+
+// ExactMatchRuleFields specifies the values for each of the fields defined in
+// an ExactMatchTable for a rule.  For example, if your ExactMatchTable was
+// configured with two fields of length 4 and 2 bytes respectively, the
+// ExactMatchRuleFields for a rule [0x89_AB_CD, 0xEF] would be
+// {{0xCD, 0xAB, 0x89}, {0xEF}} (the code assumes a little endian architecture).
+typedef std::vector<std::vector<uint8_t>> ExactMatchRuleFields;
+
+// ExactMatchTable operates as a sort-of extended CuckooMap.
+// It allows you to map multiple fields (e.g., packet headers), to some type T
+// (e.g., a gate index).
+template <typename T>
+class ExactMatchTable {
+ public:
+  using EmTable =
+      CuckooMap<ExactMatchKey, T, ExactMatchKeyHash, ExactMatchKeyEq>;
+
+  ExactMatchTable()
+      : raw_key_size_(),
+        total_key_size_(),
+        num_fields_(),
+        fields_(),
+        table_() {}
+
+  // Add a new rule.
+  //
+  // @param fields
+  //  The field values to match on (vector of bytestrings).
+  // @param val
+  //  The value to associate with packets that match this rule.
+  // Returns 0 on success, non-zero errno on failure.
+  Error AddRule(const T &val, const ExactMatchRuleFields &fields) {
+    ExactMatchKey key;
+    Error err;
+
+    if (fields.size() == 0) {
+      return MakeError(EINVAL, "rule has no fields");
+    }
+
+    if ((err = gather_key(fields, &key)).first != 0) {
+      return err;
+    }
+
+    table_.Insert(key, val, ExactMatchKeyHash(total_key_size_),
+                  ExactMatchKeyEq(total_key_size_));
+
+    return MakeError(0, std::string(""));
```

Comment:
Just cosmetics, but `MakeError(0, "");` should work fine here. Alternatively, default parameter like this..? `MakeError(int code, const std::string &msg = "")`